### PR TITLE
Fix potential crash in boats mod

### LINKED ISSUE #2342: Potential crash in boats mod
`mods/boats/init.lua`:

```
                 minetest.after(0.2, function()
                         player_api.set_animation(clicker, "sit" , 30)
                 end)
```
If `clicker` is gone, server goes BOOM.

Version 5.0.0.

### PR DESCRIPTION
**Pull request by misterskullz**
Fixes #2342 

Changes clicker to a player name string and then re-looks it up in "after" functions.  Also added in a null check on object remove, just in case.

### CODE REVIEW

**Comment by paramat** [comment]
HybridDog's comment from the other PR:
"I doubt that the nil check in after works. clicker is a local variable because it's an argument of the on_rightclick function. As far as I know, there's nothing which sets it to nil at any point in time and if a method of clicker is invoked after the player which clicker points to disappears, minetest behaves undefined, am I wrong?"

**Comment by paramat** [comment]
Has this been tested by setting the 'after' delay to a long time?

**Comment by HybridDog** [comment]
https://rubenwardy.com/minetest_modding_book/en/quality/common_mistakes.html#never-store-objectrefs-ie-players-or-entities

**Comment by paramat** [comment]
I don't think it does store objectrefs? It uses 'name'.

**Comment by HybridDog** [comment]
Sorry, I didn't look at the other commits when I wrote the comment.
https://github.com/minetest/minetest_game/pull/2351/commits
The problem has already been discussed at https://github.com/minetest/minetest_game/pull/2347#issuecomment-476954469.

**Comment by paramat** [comment]
@misterskullz has the bug been confirmed and the fix tested by setting the 'after' delay to a long time?

File: mods/boats/init.lua [hunk 1]
```diff
@@ -28,6 +28,29 @@ local function get_v(v)
...
+local function after_attach(name)
+	local player = minetest.get_player_by_name(name)
+	if player then
+		player_api.set_animation(player, "sit" , 30)
+	end
+end
+
+
+local function after_remove(object)
+	if object then
```
Comment (SmallJoker) [inline mods/boats/init.lua]: `object` will never be `nil`, because this reference is passed as a copy to `after_remove`.

**Comment by paramat** [comment]
No response, needs adoption.

File: mods/boats/init.lua (hunk 1 above)
Comment (paramat) [reply mods/boats/init.lua]: I agree, this is not needed.

File: mods/boats/init.lua [hunk 2]
```diff
@@ -133,9 +152,7 @@ function boat.on_punch(self, puncher)
 			end
 		end
 		-- delay remove to ensure player is detached
-		minetest.after(0.1, function()
-			self.object:remove()
-		end)
+		minetest.after(0.1, after_remove, self.object)
```
Comment (paramat) [inline mods/boats/init.lua]: Not needed.

**Comment by paramat** [comment]
Before we add complexity in this way i want to check the need for the delays.

The 'detach' delay was added in #630 with no reason given and i can't see any reason, perhaps we can remove it?
When detaching the player only moves upwards by 0.2 nodes so animation synchronisation seems unimportant.

The 'attach' delay was in the originally added boats mod, so perhaps has more justification, i suspect it is an animation synchronisation thing.

The delay in 'on punch' seems essential: "-- delay remove to ensure player is detached".

**Comment by paramat** [comment]
PR probably not needed, see my testing in the issue.

**Comment by HybridDog** [comment]
> PR probably not needed, see my testing in the issue.

https://github.com/minetest/minetest_game/pull/2347#issuecomment-476965778

File: mods/boats/init.lua (hunk 1 above)
Comment (HybridDog) [reply mods/boats/init.lua]: The reference is somehow set to `nil`: https://github.com/minetest/minetest_game/pull/2347#issuecomment-476965778

**Pull request closed without merging**